### DISCUSSION TITLE
Avoiding routeMessage for sending multiple messages

### DISCUSSION THREAD

**Discussion by elavy-harris** [nextgenhealthcare/connect#5651] [Q&A]
I've seen a number of posts that talk about sending multiple messages to another channel based on a single source message. They seem to talk about splitting the original into pieces and using `router.routeMessage` or `router.routeMessageByChannelId` to send the pieces. A key piece of that, (mostly?) not mentioned in the other posts, is declaring the source channel as dependent on the target channel.

I've taken a different approach. Instead of splitting up the source message, I declare the target channel to "Process Batch". The source channel then sends a single message to the target (with a Channel Writer destination, which can queue if the target is undeployed) and the target splits up the big message into pieces, each of which is processed through the target as a separate message.

Are there any particular pros/cons with either approach that I missed?

**Comment by pacmano1** [comment]
I mostly do the former - iterate over some object in the source channel and send one message at a time to the destination channnel, the destination channel does have a source queue and typically ACKs right away unless I really need the response that downstream channel.   

That is mostly about being able to "inject" a single message as needed in the destination channel from other places but also collect a response as needed from that single message.   E.g. the destination channel might go collect something on a specific patient and reply upstream that "this patient is OK" to continue on the upstream channel.   This is rather common on some order based API calls whereby a prereq must be met on your side but you gotta go ask the other side for it.

As for pros/cons, I look forward to what others say!  My stuff is driven by need rather than pro/con.

**Reply by dfauraz** [reply]
I used both approaches and got problems with the batch splitter. Now I use the routeMessage approach, even if it is annoying not to see the routed messages in the producing channel.

**Reply by dfauraz** [reply]
The batch splitter was removing some lines, an uggly bug. The key take away is : with routeMessage you do not depend on another (potentially buggy) feature to split your messages

**Reply by elavy-harris** [reply]
Any rhyme or reason to the removal? Might a JavaScript splitter have worked around that?

**Reply by dfauraz** [reply]
Yes @elavy-harris , when I say I use the "routeMessage approach", it's JavaScript

**Reply by elavy-harris** [reply]
I was asking if you used a JavaScript splitter or if you used the built in "split on MSH" splitter. Also if you have any additional information about the situations where the splitter dropped lines.

**Reply by dfauraz** [reply]
I used both and indeed the "MSH splitter" was dropping some base64 encoded lines, sometimes I don't know exactly why (there was no "MSH" in the line or around). That's why I came back to split myself with Javascript, and use routeMessage to send the messages to the destination channel. This has a few drawbacks though: you do not see the message sent in the dashboard, and you have to serialize yourself (in JavaScript).

**Reply by elavy-harris** [reply]
By any chance, do you have (and can you share) an example of a message that did not split properly? Did you create an Issue on this forum?

**Reply by dfauraz** [reply]
Nope, sorry, this is clinical data and I don't have it any more anyway
